### PR TITLE
[backport] [FLINK-6514] [build] Create a proper separate Hadoop uber jar for 'flink-dist' assembly

### PR DESCRIPTION
**Pull request by StephanEwen**
Backport of #3876 to `release-1.3`

This fixes the issue that Flink cannot be started locally if built with Maven 3.3+

There are two big fixes in this pull request, because they do not build/pass tests individually. The wrong Mesos dependencies where the reason that the broken Hadoop fat jar building actually passed the Yarn tests.

# Hadoop Uber Jar

  - This builds a proper Hadoop Uber Jar with all of Hadoop's needed dependencies. The prior build was missing many important dependencies in the Hadoop Uber Jar.

  - The Hadoop-jar is no longer excluded in `flink-dist` via setting the dependency to `provided`, but by explicit exclusion. That way, Hadoop's transitive dependencies are not excluded from other dependencies as well. Before this patch, various decompression and Avro were broken in a Flink build, due to accidental exclusion of their transitive dependencies.

# Dependency fixing

  - This also fixes the dependencies of `flink-mesos`, which made all of Hadoop's transitive dependencies its own dependencies, by promoting them during shading. That way, Flink had various unnecessary dependencies in its `flink-dist` jar.

  - Incidentally, that brought Hadoop's accidentally excluded dependencies back in, but into the `flink-dist` jar, not the `shaded-hadoop2` jar.

### CODE REVIEW

File: flink-shaded-hadoop/flink-shaded-hadoop2-uber/pom.xml [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+		<relativePath>..</relativePath>
+	</parent>
+
+	<artifactId>flink-shaded-hadoop2-uber</artifactId>
+	<name>flink-shaded-hadoop2-uber</name>
+
+	<packaging>jar</packaging>
+
+	<!--
+		the only dependency if the 'flink-shaded-hadoop2' artifact, out
```
Comment (aljoscha) [inline flink-shaded-hadoop/flink-shaded-hadoop2-uber/pom.xml]: Typo: "dependency is the ..."

**Comment by aljoscha** [comment]
The only diff from #3876 is the dropping of the jersey/jettison/servlet deps, right?

File: flink-shaded-hadoop/flink-shaded-hadoop2-uber/pom.xml [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
+	<modelVersion>4.0.0</modelVersion>
+
+	<parent>
+		<groupId>org.apache.flink</groupId>
+		<artifactId>flink-shaded-hadoop</artifactId>
+		<version>1.4-SNAPSHOT</version>
```
Comment (rmetzger) [inline flink-shaded-hadoop/flink-shaded-hadoop2-uber/pom.xml]: This needs to be 1.3-SNAPSHOT.

**Comment by rmetzger** [comment]
Merging this change ...

**Comment by rmetzger** [comment]
@StephanEwen can you manually close the PR?
Thx.

**Pull request closed without merging**